### PR TITLE
fat binary gem for Linux

### PR DESCRIPTION
**Pull request by larskanis**
With the binary linux gem it's no longer necessary to install PostgreSQL client library or header files (like package `libpq-dev`).

Solves #480

### CODE REVIEW

**Comment by larskanis** [comment]
This PR currently fails with a segfault on my local machine when connecting by a TLS encrypted connection like so:
```sh
valgrind ruby -rpg -e "PG.connect host:'localhost'"
==513704== Invalid write of size 8
==513704==    at 0x2285E800: pgtls_close (in /home/lars/comcard/ruby-pg/ports/x86_64-linux-gnu/lib/libpq.so.5.16)
==513704==    by 0x2285799F: pqsecure_close (in /home/lars/comcard/ruby-pg/ports/x86_64-linux-gnu/lib/libpq.so.5.16)
==513704==    by 0x2283E7B2: pqDropConnection (in /home/lars/comcard/ruby-pg/ports/x86_64-linux-gnu/lib/libpq.so.5.16)
==513704==    by 0x228423E0: PQconnectPoll (in /home/lars/comcard/ruby-pg/ports/x86_64-linux-gnu/lib/libpq.so.5.16)
==513704==    by 0x2277A7EE: ??? (in /home/lars/.rvm/gems/ruby-3.3.0/gems/pg-1.5.4-x86_64-linux-gnu/lib/3.3/pg_ext.so)
==513704==    by 0x4AC5FB7: rb_nogvl (thread.c:1530)
==513704==    by 0x2277AD2D: gvl_PQconnectPoll (in /home/lars/.rvm/gems/ruby-3.3.0/gems/pg-1.5.4-x86_64-linux-gnu/lib/3.3/pg_ext.so)
==513704==    by 0x227833F5: ??? (in /home/lars/.rvm/gems/ruby-3.3.0/gems/pg-1.5.4-x86_64-linux-gnu/lib/3.3/pg_ext.so)
==513704==    by 0x4B18FC3: vm_call_cfunc_with_frame_ (vm_insnhelper.c:3490)
==513704==    by 0x4B1A3AA: vm_call_symbol (vm_insnhelper.c:3953)
==513704==    by 0x4AFD436: vm_sendish (vm_insnhelper.c:5581)
==513704==    by 0x4AFD436: vm_exec_core (insns.def:834)
==513704==    by 0x4AF9CDC: rb_vm_exec (vm.c:2486)
==513704==  Address 0x222d9f60 is 1,008 bytes inside a block of size 1,192 free'd
==513704==    at 0x484810F: free (in /usr/libexec/valgrind/vgpreload_memcheck-amd64-linux.so)
==513704==    by 0x22A86323: CRYPTO_free (in /home/lars/comcard/ruby-pg/ports/x86_64-linux-gnu/lib/libpq.so.5.16)
==513704==    by 0x22A121B8: sock_free (in /home/lars/comcard/ruby-pg/ports/x86_64-linux-gnu/lib/libpq.so.5.16)
==513704==    by 0x22A05F2F: BIO_free (in /home/lars/comcard/ruby-pg/ports/x86_64-linux-gnu/lib/libpq.so.5.16)
==513704==    by 0x22A0791F: BIO_free_all (in /home/lars/comcard/ruby-pg/ports/x86_64-linux-gnu/lib/libpq.so.5.16)
==513704==    by 0x2295A79D: ossl_ssl_connection_free (in /home/lars/comcard/ruby-pg/ports/x86_64-linux-gnu/lib/libpq.so.5.16)
==513704==    by 0x2295A249: SSL_free (in /home/lars/comcard/ruby-pg/ports/x86_64-linux-gnu/lib/libpq.so.5.16)
==513704==    by 0x2285E7FB: pgtls_close (in /home/lars/comcard/ruby-pg/ports/x86_64-linux-gnu/lib/libpq.so.5.16)
==513704==    by 0x2285799F: pqsecure_close (in /home/lars/comcard/ruby-pg/ports/x86_64-linux-gnu/lib/libpq.so.5.16)
==513704==    by 0x2283E7B2: pqDropConnection (in /home/lars/comcard/ruby-pg/ports/x86_64-linux-gnu/lib/libpq.so.5.16)
==513704==    by 0x228423E0: PQconnectPoll (in /home/lars/comcard/ruby-pg/ports/x86_64-linux-gnu/lib/libpq.so.5.16)
==513704==    by 0x2277A7EE: ??? (in /home/lars/.rvm/gems/ruby-3.3.0/gems/pg-1.5.4-x86_64-linux-gnu/lib/3.3/pg_ext.so)
==513704==  Block was alloc'd at
==513704==    at 0x4845828: malloc (in /usr/libexec/valgrind/vgpreload_memcheck-amd64-linux.so)
==513704==    by 0x228443E6: makeEmptyPGconn (in /home/lars/comcard/ruby-pg/ports/x86_64-linux-gnu/lib/libpq.so.5.16)
==513704==    by 0x2283EE00: PQconnectStart (in /home/lars/comcard/ruby-pg/ports/x86_64-linux-gnu/lib/libpq.so.5.16)
==513704==    by 0x2277A7CB: ??? (in /home/lars/.rvm/gems/ruby-3.3.0/gems/pg-1.5.4-x86_64-linux-gnu/lib/3.3/pg_ext.so)
==513704==    by 0x4AC5FB7: rb_nogvl (thread.c:1530)
==513704==    by 0x2277ACD5: gvl_PQconnectStart (in /home/lars/.rvm/gems/ruby-3.3.0/gems/pg-1.5.4-x86_64-linux-gnu/lib/3.3/pg_ext.so)
==513704==    by 0x22780B76: ??? (in /home/lars/.rvm/gems/ruby-3.3.0/gems/pg-1.5.4-x86_64-linux-gnu/lib/3.3/pg_ext.so)
==513704==    by 0x4B18FC3: vm_call_cfunc_with_frame_ (vm_insnhelper.c:3490)
==513704==    by 0x4AFD436: vm_sendish (vm_insnhelper.c:5581)
==513704==    by 0x4AFD436: vm_exec_core (insns.def:834)
==513704==    by 0x4AF9CDC: rb_vm_exec (vm.c:2486)
==513704==    by 0x4968A52: rb_ec_exec_node (eval.c:287)
==513704==    by 0x4968A52: ruby_run_node (???:328)
==513704==    by 0x109204: rb_main (main.c:39)
==513704==    by 0x109204: main (main.c:58)
==513704== 
```

File: Rakefile [hunk 1]
```diff
@@ -50,24 +60,54 @@ Rake::ExtensionTask.new do |ext|
...
-				"--with-pg-lib=#{lib.static_postgresql_libdir}",
-				# libpq-fe.h resides in src/interfaces/libpq/ before make install
-				"--with-opt-include=#{lib.static_postgresql_libdir}",
+				"--with-openssl-platform=#{xlib.openssl_config}",
+				"--with-toolchain=#{xlib.toolchain}",
 			]
 		}
 	end
 
 	# Add libpq.dll to windows binary gemspec
```
Comment (cbandy) [inline Rakefile:75]: 📝 Linux and Windows, now.

Comment (larskanis) [reply Rakefile]: Thank you! Fixed.

**Comment by larskanis** [comment]
The OpenSSL issue above is resolved by #598. CI on Fedora and Alpine Linux is running and succeeds now.

With the binary linux gem it's no longer necessary to install PostgreSQL client library or header files (like package `libpq-dev`). To use one and the same libpq file named `libpq-ruby-pg.so.1` for all ruby versions saves space and the mechanism to use a `$ORIGIN`-relative `rpath` to locate and load it seems solid. The dedicated so name avoids mistakenly loaded system libpq files.

The `x86_64-linux` gem runs on Alpine Linux, but requires the package `gcompat` there as long as we don't provide a native gem for platform `x86_64-linux-musl`. It is a very small package and it seems quite common, since `nokogiri` needs it too.

So after more than one year I think this is ready to be merged.

**Pull request merged**